### PR TITLE
feat: Setup Preset Tuning Util Functions and miscellaneous validation/logging - Part 7

### PR DESCRIPTION
**Pull request by ishaansehgal99**
**Reason for Change**:
This PR introduces some of the util functions that are to be used by the function

func CreatePresetTuning(ctx context.Context, workspaceObj *kaitov1alpha1.Workspace,
	tuningObj *model.PresetParam, kubeClient client.Client) (client.Object, error)
	

This function will be in charge of handling launching the training job using the dataset source and uploading results to dataset destination. It is also incharge of ensuring configmap tuning parameters are in the right namespace so they are ready to be used by the fine_tuning_api.py file.

### CODE REVIEW

File: charts/kaito/workspace/templates/clusterrole.yaml [hunk 1]
```diff
@@ -23,7 +23,7 @@ rules:
     verbs: ["get","list","watch","create", "update", "patch" ]
   - apiGroups: [ "" ]
     resources: [ "configmaps" ]
-    verbs: [ "get","list","watch" ]
+    verbs: [ "get","list","watch","create", "delete" ]
```
Comment (ishaansehgal99) [inline charts/kaito/workspace/templates/clusterrole.yaml:26]: Needed to now create duplicate configmap from release to workspace namespace

File: pkg/tuning/preset-tuning.go [hunk 2]
```diff
@@ -2,11 +2,142 @@ package tuning
...
+	registryName := os.Getenv("PRESET_REGISTRY_NAME")
+	return formatRegistryImagePath(registryName, "kaito-tuning-"+string(wObj.Tuning.Preset.Name), presetObj.Tag)
+}
+
+func GetDataSrcImageInfo(ctx context.Context, wObj *kaitov1alpha1.Workspace) (string, []corev1.LocalObjectReference) {
+	imagePullSecretRefs := make([]corev1.LocalObjectReference, len(wObj.Tuning.Input.ImagePullSecrets))
+	for i, secretName := range wObj.Tuning.Input.ImagePullSecrets {
+		imagePullSecretRefs[i] = corev1.LocalObjectReference{Name: secretName}
+	}
+	registryName := os.Getenv("PRESET_REGISTRY_NAME")
```
Comment (Fei-Guo) [inline pkg/tuning/preset-tuning.go:92]: Why do we need to get registry name from env?  Can we get the registry from the Image, 
or in the other word, wObj.Tuning.Input.Image should be a valid image url already.

File: pkg/tuning/preset-tuning.go [hunk 3]
```diff
@@ -2,11 +2,142 @@ package tuning
...
+}
+
+func GetDataDestImageInfo(ctx context.Context, wObj *kaitov1alpha1.Workspace) (string, []corev1.LocalObjectReference) {
+	registryName := os.Getenv("PRESET_REGISTRY_NAME")
+	imageName := fmt.Sprintf("%s/%s", registryName, wObj.Tuning.Output.Image)
+	imagePushSecretRefs := []corev1.LocalObjectReference{{Name: wObj.Tuning.Output.ImagePushSecret}}
+	return imageName, imagePushSecretRefs
+}
+
+func CreatePresetConfigMap(ctx context.Context, workspaceObj *kaitov1alpha1.Workspace,
```
Comment (Fei-Guo) [inline pkg/tuning/preset-tuning.go:104]: Usually, if we check if the configmap exists before creating a new one, we call this method "EnsurePresetConfigmap".

How about EnsureTuningConfigMap?

**Review by Fei-Guo — COMMENTED** [review]
remove deadcode.

File: pkg/tuning/preset-tuning.go (hunk 2 above)
Comment (ishaansehgal99) [reply pkg/tuning/preset-tuning.go]: Ok - User will specify full url

File: pkg/tuning/preset-tuning.go (hunk 3 above)
Comment (ishaansehgal99) [reply pkg/tuning/preset-tuning.go]: Changed

File: api/v1alpha1/workspace_validation.go [hunk 4]
```diff
@@ -163,6 +163,12 @@ func (r *DataSource) validateCreate() (errs *apis.FieldError) {
 		sourcesSpecified++
 	}
 	if r.Image != "" {
+		if !strings.Contains(r.Image, ":") {
```
[the remark this answers is not in the document]
Comment (ishaansehgal99) [reply api/v1alpha1/workspace_validation.go]: added regex

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**